neps/nep-0539.md

```diff
@@ -0,0 +1,741 @@
+
+    - Problem 1: Without knowing which shards a transaction touches and how
+      large receipts will be, we have to pessimistically reserve more space than
+      most receipts will actually need.
+    - Problem 2: If buffer space is shared globally, individual queues can still
+      grow really large, even indefinitely if we assume the number of shards
+      grows over time.
+    - Problem 3: If buffer space is on a per-shard basis, we run into deadlocks
+      when two shards have no more space left but both need to send to the other
+      shard to make progress.
+
+6. Require users to define which shards their transactions will touch and how
+   much gas is burnt in each. Then use this information for global scheduling
+   such that congestion is impossible.
+
+    - Problem 1: This requires lots of changes across the infrastructure stack.
+      It would take too long to implement as we are already facing congestion
+      problems today.
+    - Problem 2: This would have a strong impact on usability and it is unclear
+      if gas usage estimating services could close the gap to make it
+      acceptable.
+
+
+
+## Future possibilities
+
+While this proposal treats all requests the same, it sets the base for a proper
+transaction priority implementation. We co-designed this proposal with
+[NEP-541](https://github.com/near/NEPs/pull/541), which adds a transaction
+priority fee. On a very high level, the fee is used to auction off a part of the
+available gas per chunk to the highest bidders.
+
+We also expect that this proposal alone will not be the final solution for
+congestion control. Rather, we just want to build a solid foundation in this NEP
+and allow future optimization to take place on top of it.
+
+For example, estimations on how much gas is burnt on each shard could help with
+better load distribution in some cases.
+
+We also forsee that the round-robin scheduling of shard allowed to forward even
+under full congestion is not perfect. It is a key feature to make deadlocks
+provably impossible, since every shard is guaranteed to make a minimum progress
+after N rounds. But it could be beneficial to allocate more bandwidth to shards
+that actually have something to forward, or perhaps it would be better to stop
+forwarding anything for a while. The current proposal allows chunk producers to
+experiment with this without a protocol version change.
+
+Lastly, a future optimization could do better transaction rejection for meta
+transactions. Instead of looking only at the outer transaction receiver, we
+could also look at the receiver of the delegate action, which is most likely
+where most gas is going to be burnt, and use this for transaction rejection.
+
+## Consequences
+
+
+### 
+
+### Positive
+
+- Accepted transaction have lower latencies compared to today under congestion.
+- Storage and memory requirements on validator for storing receipts are bounded.
```

Comment:
I don't understand why outgoing receipts buffers are bounded in size. Or at least I only understand why this is a case under the assumption that `filter_tx` above is working effectively. Otherwise, I can imagine a case when we keep accepting and processing transactions to shard X (due to a low value of `incoming_congestion`) that generate an outgoing receipt (as a cross-contract call) to a congested shard Y which can't all be processed timely and accumulate in unbounded fashion.

Reply:
Absolutely right, outgoing buffers are not hard-bounded to the memory size parameter we give it. We should probably clarify what we mean by a bound.

In simulations, we tried to push it past the soft-limit as much as possible. One example is in the spreadsheet. You can see 1.3 GB memory usage for the "Big Linear Imbalance" workload, with the limit at 1GB. This workload has 500 B / 60 Tgas receipts that indirectly produce 2 MB / 10 Tgas receipts.

But we tried with many other workloads, too. Every workload will fill up the system with some amount of receipts before it locks down new traffic from entering. And max memory usage then essentially depends on what memory the receipts in the system use to finish enough (child) receipts to allow more traffic to enter again.

This arguing only works if the system actually locks down every workload eventually. This is hard, if not impossible, to prove for general dynamic workloads. That's why we decided to simulate as much as we could.

In practice, we can still cause the memory requirements to 3-4x the limit. But all simulated workloads stabilized around that or lower. In comparison, today's "strategy" would lead to >500GB memory usage within 2000 blocks even for the aforementioned "Big Linear Imbalance" workload.

All that said, we are aware that for malicious workloads, the max in bytes could still be larger than we can handle. There are a couple malicious ideas we never tested. But from the beginning, we put malicious workloads as second priority behind getting a basic congestion control for regular users in place.